llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp

```diff
@@ -473,13 +589,17 @@ void NVPTXLowerArgs::handleByValParam(const NVPTXTargetMachine &TM,
     return;
   }
 
-  const DataLayout &DL = Func->getDataLayout();
+  // We can't access byval arg directly and need a pointer. on sm_70+ we have
+  // ability to take a pointer to the argument without making a local copy.
+  // However, we're still not allowed to write to it. If the user specified
+  // `__grid_constant__` for the argument, we'll consider escaped pointer as
+  // read-only.
   unsigned AS = DL.getAllocaAddrSpace();
-  if (isParamGridConstant(*Arg)) {
-    // Writes to a grid constant are undefined behaviour. We do not need a
-    // temporary copy. When a pointer might have escaped, conservatively replace
-    // all of its uses (which might include a device function call) with a cast
-    // to the generic address space.
+  if (HasCvtaParam && (!(PI.isEscaped() || PI.isAborted()) || IsGridConstant)) {
+    LLVM_DEBUG(dbgs() << "Using non-copy pointer to " << *Arg << "\n");
+    // Replace all argument pointer uses (which might include a device function
```

Comment:
Just for my understanding- for device function calls for non-grid constant parameters, PI.isEscaped will be true, due to https://github.com/llvm/llvm-project/blob/552e5c3e6b906d71976fc0f1205f79d636039b6c/llvm/include/llvm/Analysis/PtrUseVisitor.h#L292, and a copy *will* be made?

Reply:
Correct, calls allow the pointers to escape. Except for the memcpy/memmove that we're handling separately.